src/test/java/integration/TestRequestManager.java

```diff
@@ -0,0 +1,425 @@
+        testLogin(user2ID, user2Pass);
+        testRemoveSelfFromProject(proj2.getProjectID());
+        testLogout();
+        testLogin(user1ID, user1Pass);
+        deleteProject(proj2.getProjectID());
+    }
+
+    private void testFetchPermissionConstants() throws InterruptedException {
+        logger.info("Fetching permission constants");
+        Semaphore waiter = new Semaphore(0);
+        PropertyChangeListener listener = (event) -> {
+            if (event.getPropertyName().equals(SessionStorage.PERMISSION_CONSTANTS)) {
+                apiConstants = (BiMap<String, Byte>) event.getNewValue();
+                waiter.release();
+            }
+        };
+        dataMgr.getSessionStorage().addPropertyChangeListener(listener);
+        reqMgr.fetchPermissionConstants();
+        if (!waiter.tryAcquire(5, TimeUnit.SECONDS)) {
+            Assert.fail("Acquire timed out");
+        }
+        dataMgr.getSessionStorage().removePropertyChangeListener(listener);
+    }
+
+    private void registerUser(String userID, String userFirstName,
+                              String userLastName, String userEmail, String userPass) throws InterruptedException, java.net.ConnectException {
+        logger.info(String.format("Registering user: %s", userID));
+        Semaphore waiter = new Semaphore(0);
+
+        req = new UserRegisterRequest(userID, userFirstName, userLastName, userEmail, userPass).getRequest(response -> {
+            // If registration fails, probably is already there.
+            waiter.release();
+        }, errHandler);
+        ws1.sendRequest(req);
+        if (!waiter.tryAcquire(5, TimeUnit.SECONDS)) {
+            Assert.fail("Acquire timed out");
+        }
+    }
+
+    private void testLogin(String userID, String userPass) throws InterruptedException {
+        logger.info("Logging in");
+        Semaphore waiter = new Semaphore(0);
+        final String[] senderToken = new String[1];
+        PropertyChangeListener listener = (event) -> {
+            if (event.getPropertyName().equals(SessionStorage.USERNAME)) {
+                if (userID.equals(event.getNewValue())) {
+                    waiter.release();
+                }
+            } else if (event.getPropertyName().equals(SessionStorage.AUTH_TOKEN)) {
+                if (event.getNewValue() != null) {
+                    senderToken[0] = sesSto.getAuthenticationToken();
+                    waiter.release();
+                }
+            }
+        };
+        sesSto.addPropertyChangeListener(listener);
+        reqMgr.login(userID, userPass);
+        if (!waiter.tryAcquire(2, 5, TimeUnit.SECONDS)) {
+            Assert.fail("Acquire timed out");
+        }
+        sesSto.removePropertyChangeListener(listener);
+    }
+
+    private Project testCreateProject(String name) throws InterruptedException {
+        logger.info("Creating project");
+        Semaphore waiter = new Semaphore(0);
+        final Project[] toReturn = new Project[1];
```

Comment:
This function is called multiple times to create different functions. I can add a field instead of this toReturn variable, but it will serve the same purpose.